### PR TITLE
Enhanced Parallel Processing and Plotting in ELMPostprocessR

### PR DESCRIPTION
**Pull request by sharma-bharat**
With this latest update, ELMPostprocessR now supports parallel execution for both processing and plotting the ensemble members of simulation outputs independently.

### CODE REVIEW

**Comment by sharma-bharat** [comment]
Examples of bash scripts to run ELM_postprocessR on single and ensemble runs.

[postprocessR_singleFACE.txt](https://github.com/walkeranthonyp/ELM_postprocessR/files/15240732/postprocessR_singleFACE.txt)
[postprocessR_ensembleFACE_parallel.txt](https://github.com/walkeranthonyp/ELM_postprocessR/files/15240733/postprocessR_ensembleFACE_parallel.txt)

File: functions_plotting.R [hunk 1]
```diff
@@ -275,24 +275,29 @@ plotlist_png <- function(plotlist, fname='plotlist_png.pdf', ... ) {
...
   print(paste('Saving figures to:',plotname), quote=F )
- 
+
   if(png) {
     plotlist_png(plots, plotname )
- 
-  } else { 
+
+  } else {
+    print ("BS...>>>")
```
Comment (walkeranthonyp) [inline functions_plotting.R]: Please remove these "BS" tags

File: plot_ELM.R [hunk 2]
```diff
@@ -23,29 +23,31 @@ plot_nutrients <- F
...
   caseidprefix   <- 'FACEconly_exptest_vcmax70'
   sites          <- 'US-DUK'
   case_labs      <- c('spins', 'trans' )
-  highfreq_plots <- F  
-  
-  
+  highfreq_plots <- F
+
+  # either NULL or an integer number of ensemble members
+  uq              <- NULL
```
Comment (walkeranthonyp) [inline plot_ELM.R]: Please align the "<-" with above

File: plot_ELM.R [hunk 3]
```diff
@@ -70,7 +72,11 @@ source('functions_plotting.R')
 # plotting lists
 source('info_plotting_fatesoutv2.R')
 
-
+nuq <- if(is.null(uq)) 1 else  uq
+# command line argument uq_index to parallelize the
+if(nuq>1){
+print(paste('Plotting for ensemble index : ',uq_index,'/', nuq))
+}
```
Comment (walkeranthonyp) [inline plot_ELM.R:79]: Please indent the print line by 2 columns, or make if all one line

File: plot_ELM.R [hunk 4]
```diff
@@ -86,80 +92,90 @@ if(is.null(wd_out_plots)) {
...
-#c <- 2
 for(c in 1:length(case_labs)) {
 
   fname   <- paste(caseidprefix,sites,case_labs[c],sep='_')
   fname_a <- paste0(fname,'_annual.RDS')
-  
+
+  # overwriting the fname and fname_a if running UQ ensembles
+  if(nuq>1){
+    fname   <- paste(caseidprefix,sites,case_labs[c],sep='_')
```
Comment (walkeranthonyp) [inline plot_ELM.R]: Please align all of the "<-" to the same column and make sure they're followed by a single space

File: process_ELM_output.R [hunk 5]
```diff
@@ -1,9 +1,9 @@
 ##############################
 #
-# Open ELM/CLM netcdf files, join, & put into lists of R arrays 
+# Open ELM/CLM netcdf files, join, & put into lists of R arrays
 #
 # AWalker
-# April 2021
```
Comment (walkeranthonyp) [inline process_ELM_output.R:6]: Please retain the initial date (April 2021)

File: process_ELM_output.R [hunk 6]
```diff
@@ -45,34 +45,36 @@ sites           <- 'US-DUK'
...
-# - currently only this one is supported, anything other then output files covering integer years requires development 
+# - currently only this one is supported, anything other then output files covering integer years requires development
 fstart_datetime <- '-01-01-00000'
 # netcdf dimensions that contain character variables, can be a vector
 char_dims       <- 'string_length'
-# the number of members in a UQ ensemble, NULL means no ensemble 
+# the number of members in a UQ ensemble, NULL means no ensemble
 uq_ensemblen    <- NULL
+# the id number member in a UQ ensemble, NULL means no ensemble
+uq_index    <- NULL
```
Comment (walkeranthonyp) [inline process_ELM_output.R]: align <-

File: process_ELM_output.R [hunk 7]
```diff
@@ -138,89 +140,105 @@ print('',quote=F)
...
-    
-    # UQ loop
-    for(u in 1:nuq) {
-    
+
+    # UQ Input arguments
+    # Ensemble Processing Update:
+    # Aim: to run the ensembles in parallel (compared to a for loop)
+    # Additional requirement: you need to pass additional arguments `uq_index` and `uq`...
+    # ... `uq_index` is the index/id of the ensemble run
```
Comment (walkeranthonyp) [inline process_ELM_output.R]: please change "run" to "member" in this comment

File: process_ELM_output.R [hunk 8]
```diff
@@ -138,89 +140,105 @@ print('',quote=F)
...
+    # ... `uq` is the number of ensembles
```
Comment (walkeranthonyp) [inline process_ELM_output.R:184]: I think you mean number of ensemble members

File: process_ELM_output.R [hunk 9]
```diff
@@ -138,89 +140,105 @@ print('',quote=F)
...
+    u <- if(nuq == 1) 1 else uq_index
+    print(uq_index)
```
Comment (walkeranthonyp) [inline process_ELM_output.R]: Please combine these print statements

**Review by walkeranthonyp — CHANGES_REQUESTED** [review]

**Comment by sharma-bharat** [comment]
I have made the changes based on your comments. Please review them!

**Review by walkeranthonyp — APPROVED** [review]
Thanks!

**Pull request merged**